### PR TITLE
More flags to parametrise generation

### PR DESCRIPTION
**Pull request by phadej**

### CODE REVIEW

File: .travis.yml [hunk 1]
```diff
@@ -123,6 +117,7 @@ install:
       cabal new-update head.hackage -v
     fi
   - grep -Ev -- '^\s*--' ${HOME}/.cabal/config | grep -Ev '^\s*$'
+  - if [ $HCNUMVER -eq 80403 ]; then cabal new-install -w ${HC} --symlink-bindir=$HOME/.local/bin hlint --constraint='hlint ==2.1.*'; fi
```
Comment (RyanGlScott) [inline .travis.yml]: One thing to keep in mind here is that building `hlint` from scratch will contribute to the build time. Are we sure we don't want to just install it from `apt-get`?

File: .travis.yml [hunk 2]
```diff
@@ -1,6 +1,6 @@
 # This Travis job script has been generated by a script via
 #
-#   runghc make_travis_yml_2.hs '-o' '.travis.yml' '--ghc-head' '--irc-channel=irc.freenode.org#haskell-lens' '--no-no-tests-no-bench' '--no-unconstrained' 'cabal.project'
+#   runghc make_travis_yml_2.hs '-o' '.travis.yml' '--ghc-head' '--irc-channel=irc.freenode.org#haskell-lens' '--no-no-tests-no-bench' '--no-unconstrained' 'cabal.project' '--hlint' '--hlint-options=-cpp-define=HLINT' '--allow-failure=7.0.4' '--allow-failure=7.2.2' '--env=7.8.4:NOINSTALLEDCONSTRAINTS=true' '--env=7.6.3:NOINSTALLEDCONSTRAINTS=true' '--env=7.4.2:NOINSTALLEDCONSTRAINTS=true' '--env=7.2.2:NOINSTALLEDCONSTRAINTS=true'
```
Comment (RyanGlScott) [inline .travis.yml]: Ooh, those `--allow-failure` and `--env` flags are nice. I'm definitely going to make heavy use of those.

File: .travis.yml (hunk 1 above)
Comment (phadej) [reply .travis.yml]: it's cached, dependencies are somewhat stable, and it's only single job (no reason to `hlint` in every matrix sub-job)

Comment (phadej) [reply .travis.yml]: ... and you get newer hlint (or actually you can control which version you get, with `--hlint-version`)

Comment (RyanGlScott) [reply .travis.yml]: Alright, you've convinced me :)

One more question about this: it looks like the script currently builds all of the dependencies for `hlint` sequentially. Can this be made parallel?

**Comment by RyanGlScott** [comment]
It looks like the only remaining build failures are suggestions from a newer version of `hlint`.

The "Use String" suggestion is probably not warranted for this package, so I'd add `ignore "Use String"` to `HLint.hs`.

I'm not sure I understand what the "Reduce duplication" suggestions mean.

**Comment by phadej** [comment]
there are code paths which looks suspiciously similar, sometimes those are valid suggestions, sometimes they aren't.

**Comment by phadej** [comment]
i.e I'd just ignore them too, can I?

**Comment by RyanGlScott** [comment]
I'd be fine with ignoring that suggestion as well, sure.

**Comment by phadej** [comment]
green.

**Pull request merged**